src/core/xds/grpc/xds_matcher_parse.cc

```diff
@@ -0,0 +1,354 @@
+    const XdsResourceType::DecodeContext& context,
+    const xds_type_matcher_v3_Matcher_MatcherList_Predicate* predicate,
+    const UniqueTypeName& matcher_context, ValidationErrors* errors);
+
+// Function to parse "xds_core_v3_TypedExtensionConfig" to generate
+// XdsMatcher::Input<T>
+// The parsing is for input which return absl::string_view
+std::unique_ptr<XdsMatcher::InputValue<absl::string_view>> ParseStringInput(
+    const XdsResourceType::DecodeContext& context,
+    const xds_core_v3_TypedExtensionConfig* input,
+    const UniqueTypeName& matcher_context, ValidationErrors* errors) {
+  const google_protobuf_Any* any =
+      xds_core_v3_TypedExtensionConfig_typed_config(input);
+  auto extension = ExtractXdsExtension(context, any, errors);
+  if (!extension.has_value()) {
+    return nullptr;
+  }
+  const auto& registry =
+      DownCast<const GrpcXdsBootstrap&>(context.client->bootstrap())
+          .matcher_string_input_registry();
+  auto result = registry.ParseAndCreateInput(context, *extension,
+                                             matcher_context, errors);
+  return result;
+}
+
+// Function to parse "xds_core_v3_TypedExtensionConfig"  to generate
+// supported Actions
+std::unique_ptr<XdsMatcher::Action> ParseAction(
+    const XdsResourceType::DecodeContext& context,
+    const xds_core_v3_TypedExtensionConfig* action,
+    const XdsMatcherActionRegistry& action_registry, ValidationErrors* errors) {
+  const google_protobuf_Any* any =
+      xds_core_v3_TypedExtensionConfig_typed_config(action);
+  auto extension = ExtractXdsExtension(context, any, errors);
+  if (!extension.has_value()) {
+    return nullptr;
+  }
+  return action_registry.ParseAndCreateAction(context, *extension, errors);
+}
+
+// Parse and generate input matcher with type string_view
+// Parsing "xds_type_matcher_v3_StringMatcher" to generate StringMatcher
+std::unique_ptr<XdsMatcherList::InputMatcher<absl::string_view>>
+ParseStringMatcher(const XdsResourceType::DecodeContext& context,
+                   const xds_type_matcher_v3_StringMatcher* string_matcher_upb,
+                   ValidationErrors* errors) {
+  auto string_matcher = StringMatcherParse(context, string_matcher_upb, errors);
+  return std::make_unique<XdsMatcherList::StringInputMatcher>(
+      std::move(string_matcher));
+}
+
+// Parse OnMatch components of the matcher
+// TODO(bpawan): b/431645556 this should have a max recusrsion depth
+std::optional<XdsMatcher::OnMatch> ParseOnMatch(
+    const XdsResourceType::DecodeContext& context,
+    const xds_type_matcher_v3_Matcher_OnMatch* on_match,
+    const XdsMatcherActionRegistry& action_registry,
+    const UniqueTypeName& matcher_context, ValidationErrors* errors) {
+  // TODO(bpawan): b/431645620 Parse keep matching once we move to latest xds
+  // protos
+  bool keep_matching = false;
+  // Action is a variant which can have Action or a Nested Matcher
+  if (xds_type_matcher_v3_Matcher_OnMatch_has_action(on_match)) {
```

Comment:
Instead of calling a "has" method here and then accessing the field later, I suggest combing them into one step -- just access the field and then check whether it's null:

```
if (const auto* action_proto = xds_type_matcher_v3_Matcher_OnMatch_action(on_match);
    action_proto != nullptr) {
  ValidationErrors::ScopedField field(errors, ".action");
  auto action = ParseAction(context, action_proto, action_registry, errors);
  return XdsMatcher::OnMatch(std::move(action), keep_matching);
}
```

Same thing throughout.